reference/test/solver/bicgstab_kernels.cpp

```diff
@@ -108,6 +163,249 @@ class Bicgstab : public ::testing::Test {
+
+    gko::kernels::reference::bicgstab::step_2(
+        this->exec, this->small_r.get(), this->small_s.get(),
+        this->small_v.get(), this->small_rho.get(), this->small_alpha.get(),
+        this->small_beta.get(), &this->small_stop);
+
+    GKO_ASSERT_MTX_NEAR(this->small_s, l({{-2.25, 5.0}, {-2.25, 5.0}}), 0);
+    GKO_ASSERT_MTX_NEAR(this->small_alpha, l({{0.25, 0.0}}), 0);
+}
+
+
+TYPED_TEST(Bicgstab, KernelStep2DivZero)
+{
+    this->small_s->fill(5);
+    this->small_r->fill(-2);
+    this->small_v->fill(1);
+    this->small_alpha->fill(4);
+    this->small_rho->fill(1);
+    this->small_beta->fill(0);
+
+    gko::kernels::reference::bicgstab::step_2(
+        this->exec, this->small_r.get(), this->small_s.get(),
+        this->small_v.get(), this->small_rho.get(), this->small_alpha.get(),
+        this->small_beta.get(), &this->small_stop);
+
+    GKO_ASSERT_MTX_NEAR(this->small_s, l({{-2.0, -2.0}, {-2.0, -2.0}}), 0);
+    GKO_ASSERT_MTX_NEAR(this->small_alpha, l({{0.0, 0.0}}), 0);
+}
+
+
+TYPED_TEST(Bicgstab, KernelStep3)
+{
+    this->small_x->fill(5);
+    this->small_r->fill(-2);
+    this->small_s->fill(1);
+    this->small_y->fill(4);
+    this->small_z->fill(-6);
+    this->small_t->fill(7);
+    this->small_omega->fill(10);
+    this->small_beta->at(0) = 2;
+    this->small_beta->at(1) = 3;
+    this->small_gamma->at(0) = 8;
+    this->small_gamma->at(1) = 3;
+    this->small_alpha->at(0) = 1;
+    this->small_alpha->at(1) = -2;
+    this->small_stop.get_data()[1] = this->stopped;
+
+    gko::kernels::reference::bicgstab::step_3(
+        this->exec, this->small_x.get(), this->small_r.get(),
+        this->small_s.get(), this->small_t.get(), this->small_y.get(),
+        this->small_z.get(), this->small_alpha.get(), this->small_beta.get(),
+        this->small_gamma.get(), this->small_omega.get(), &this->small_stop);
+
+    GKO_ASSERT_MTX_NEAR(this->small_x, l({{-15.0, 5.0}, {-15.0, 5.0}}), 0);
+    GKO_ASSERT_MTX_NEAR(this->small_r, l({{-27.0, -2.0}, {-27.0, -2.0}}), 0);
+    GKO_ASSERT_MTX_NEAR(this->small_omega, l({{4.0, 10.0}}), 0);
+}
+
+
+TYPED_TEST(Bicgstab, KernelStep3DivZero)
+{
+    this->small_x->fill(5);
+    this->small_r->fill(-2);
+    this->small_s->fill(1);
+    this->small_y->fill(4);
+    this->small_z->fill(-6);
+    this->small_t->fill(7);
+    this->small_omega->fill(10);
+    this->small_beta->fill(0);
+    this->small_gamma->at(0) = 8;
+    this->small_gamma->at(1) = 3;
+    this->small_alpha->at(0) = 1;
+    this->small_alpha->at(1) = -2;
+
+    gko::kernels::reference::bicgstab::step_3(
+        this->exec, this->small_x.get(), this->small_r.get(),
+        this->small_s.get(), this->small_t.get(), this->small_y.get(),
+        this->small_z.get(), this->small_alpha.get(), this->small_beta.get(),
+        this->small_gamma.get(), this->small_omega.get(), &this->small_stop);
+
+    GKO_ASSERT_MTX_NEAR(this->small_x, l({{9.0, -3.0}, {9.0, -3.0}}), 0);
```

Comment:
I think the check for `this->small_r` is missing. It should be
```suggestion
    GKO_ASSERT_MTX_NEAR(this->small_x, l({{9.0, -3.0}, {9.0, -3.0}}), 0);
    GKO_ASSERT_MTX_NEAR(this->small_r, this->small_s, 0);
```